### DISCUSSION TITLE
Pulsar's "thing"

### DISCUSSION THREAD

**Discussion by Daeraxa** [pulsar-edit/.github#27] [General]
Just want to start a conversation as to what Pulsar's "thing" is that will help it stand out or at least attract certain users. Obviously we have to take into account what is out there and what is upcoming to compare with.

"Hackable" is clearly the most obvious part of Pulsar's identity but what does that really mean? It feels like Atom didn't really go far enough with it and from conversations on Discord it does sound like this could be massively emphasised by extending the provider/services model it already has.

But is this the only area we want to target? Obviously we bring Atom's legacy along with us as its successor but I think there is opportunity here to try and stand out.
For example maybe it should not only should "hackable" but "easily hackable" with a lower barrier of entry, high quality documentation/tutorials etc. to really make it absurdly customisable to most users both entry level and advanced.
Obviously I think it should still provide a good "out of box" experience as well, many people may well be picking it up as their first editor for any number of reasons so poor language support and a complicated mess of packages to get basic functionality wouldn't be a good first impression.

To try and explain what I mean I've done a short analysis of the most common points of comparison below and what tends to make them stick out from the others.

### VSCode
My take on VSCode's "thing" is that it is a Swiss army knife. It does extremely well "out of the box" for all kinds of popular languages and comes with all the common accessories that people want - terminal, git, debugger etc. but also has a ton of extensions to make it go further.
In fact the number of extension might just be the thing that makes it - 40k of them and with openVSX they might well just become a standard.
Also has a web version so it isn't even limited to being installed.

### Brackets/Phoneix
Brackets/Phoenix I think is spiritually very similar to Pulsar, both are continuations of projects abandoned by its creators. Brackets really focuses hard on the web development side of things, it comes with a website live preview and, whilst extensible, really seems to focus hard on the inbuilt stuff and being generally pretty lightweight.
Like VSCode it has a web based version (Phoenix) which fits it well thematically.

### Zed
Zed is obviously a newcomer, made by much of the old Atom team and its "thing" is that it is incredibly fast - no javascript, no electron etc, all written in rust (and it is *fast*, significantly quicker than Atom on a very slow mac mini). They are very big on pushing the collaborative editing side - very much like how Teletype was so prominently featured with Atom.
Like Pulsar they are also pushing for the Atom fans as a bit of a "spiritual successor" to Atom and they can do with Zed what they couldn't do with Atom.
Not currently open source but it is planned to be once it leaves early development.

### Lapce
Lapce doesn't seem to have that much identity yet. It pushes itself, like Zed, as a responsive and very fast, rust-based editor. The other bits they push are being able to do "remote" development and their WASI plugin system.

### Sublime Text
Not open source but definitely in the same arena, really pitches itself more like VSCode as a generally competent editor with a great out of the box experience. Used to really be a stand out product in a market of very little real competition.

**Comment by kaosine** [comment]
I mean, definitely don't want to be a jack of all trades right out of the gate and defeat ourselves with too much work. I'd just lean further into the hackable nature and do what we can to out of the box make things easy for people to jump in and customize everything they want. Especially if we look at what they planned for the Chapters in the docs after the like 5 they did before you get to the appendix stuff?

That could be a partial way forward once we're done modernizing the backend dependencies that run the core editor. Since it definitely seemed to be the plan before M$ crashed the party....

**Comment by confused-Techie** [comment]
Personally I'd like to see Pulsar go further into doing what developers need, saving use time wherever possible. Now this could somewhat take place from integrations. Of course of which should be optional packages. So along with extreme hackability, I think we need to have a different set of 1st party packages created. 

Now we have a lot on our plate right now, so this may be for later, but let me get to some examples of what I mean.

I think we should find what can save developers time in their respective field and incorporate it as well as possible. Now obviously this can mean more Git integration, or specifically more GitHub/Bitbucket integration. More terminal integration, but maybe even integration with common 3rd party tools. Like Codacy integration, Coveralls, and so on. Get what the developer needs in one place. 

Now here are some things I feel like are of the same idea, but I've always wanted to see in the editor, just never got around to making them.

* Automatic config creation, as in a way to generate config files within the editor, like running `npm init -y` or `go mod init` and the like.
* Documentation Integration: I've always thought it would be the coolest thing if I typed `Array.map...` and hovering on it could give a short tidbit about that function, with clicking on it opening a pane within the editor showing MDN's docs on it. 
* Continuing on Documentation Support, even just pulling up the relevant docs in a sidebar depending on an autodetected usage of what language and service you're using. Like you have a `package.json` ok heres some NodeJS Docs, ah this file is importing `express` here's some ExpressJS docs.
* And lastly more previews. We can already preview MarkDown in the editor, which is extremely helpful while writing a `README.md` for example, but imagine we could even preview things like the site we are building with a panel that loads it up so we can see it while working on it or something. 

I know theres some kinda random ideas here, that should absolutely be intended as packages rather than core editor features, just thought I'd take this moment to get some of these ideas solidified from my head.

But what I'm trying to say, is I'd like to see Pulsars thing be:
1) A extremely hackable text editor
2) An editor built for developers. It helps you not only write your code, but can help you find the docs for it, can help you preview or visualize it. And even help you configure it, or at least save the mundane steps of development within a certain space. Thats what I'd really like to see us lean into. But I know thats a lot.

**Reply by mauricioszabo** [reply]
Well, most of these can (and some of these already have) be implemented with packages. Maybe we can add more documentation on that, but one of the wonderful things about the Atom/Pulsar API is how **insanely flexible** it is - we can change almost any aspect of the editor, add incredible decorations that are complete HTML+CSS+Javascript things on their own, hook up the whole NPM ecosystem for that (even binary packages) on **any stage**, we can even crash the editor completely with a package (which is dangerous, but gives an immense power).

What we may need to do is either triage some wonderful packages and offer it as recommendations, or pitch how easy is to do packages on Pulsar and wait for good packages to appear (maybe this option is not feasible, I don't know).

I'm actually going to open a different "answer" on this discussion showing my exp. with writing a plug-in too.

**Comment by mauricioszabo** [comment]
Pulsar, for me, is a modern Emacs. It can do _almost everything_ that Emacs can, down to an extent, and it does really well. What I mean is:

On Emacs, you can configure almost any place of the editor. Want to move the cursor 2 lines down, 3 lines left, and type "Hello, world"? Well, you can, with some commands. Pulsar allows pretty much the same. But Pulsar is also more "visual" and allows an user to configure on technology that most people know (HTML, CSS, Javascript, not a specific language only used inside the editor) and also allows for very interesting situations like using NPM modules, even native ones, to write plug-ins or even inside the init script.

Now, to show the power of the editor: VSCode is also visual, HTML+CSS+JS, but you can't actually _modify_ the editor in any way, unless there's an API for it. You can generate a "webview" that lives inside a tab/split and arbitrary HTML+CSS+JS there, but you can't call NPM modules so your choices are limited (sure, you can pipe commands through the editor and call NPM there, but again, limited).

As an example of what Pulsar is capable of, I have a project called Chlorine - it allows the editor to connect to a socket that represents a live Clojure project running elsewhere. It can call commands on it, and it can visualize the results graphically. My Pulsar is also running on VIM mode, and I am one of the people that keep forgetting if I'm on "Normal mode" or on "Insert mode". So:

When I'm on edit mode, my background is one, and on insert mode, my background is other.
Also, I like ligatures, but editing then are SUCH A PAIN, so when I'm on insert mode, I disable it
Or, when I'm on the line that ligatures are present.
Also, on Clojure code lines have no meaning, so I disable the line highlight.
![Customization-optimized](https://user-images.githubusercontent.com/138037/193499547-dab6152a-213b-4cb6-9f64-44c91bbc88ae.gif)

Now, Chlorine: I can render the result of an evaluation, and it gets rendered as HTML. So I can interact with it, as if it was a "tree" and not only as text:
![rendering-1](https://user-images.githubusercontent.com/138037/193499702-19b7884e-90e9-4e72-aabc-31be370687dc.gif)

And I can do that because it's pure HTML and Javascript. But hey, if this exist, why not render something and immediately make the content full screen? Also, I could click on the content and change it, like a... slideshow? A poor man's powerpoint if you will?
![slideshow](https://user-images.githubusercontent.com/138037/193499856-e71f70bb-1d9f-4f3a-a49e-6e0b938a4de3.gif)

Also, if I can render arbitrary HTML, why should I be limited on what the _language returned_ to me? Why can't I use that result and generate an HTML that makes more sense for me?
![Moves](https://user-images.githubusercontent.com/138037/193500137-51f89aa4-6f9b-48b8-a306-ba50d775e67f.gif)

And obviously, you can also configure _that_ to accept clicks and such:
![Moves 2](https://user-images.githubusercontent.com/138037/193500183-75497384-2d31-4cb2-86e0-e212a98dd8d7.gif)

Making a full "code visualizer" that actually _helps you_ visualize what you're doing, instead of just a just "glorified print" that will help, sure, but not enough.

And why is it possible? Because everything is Javascript - the plug-in runs arbitrary Javascript, and allows the user to customize things with arbitrary Javascript. The user can customize the editor with HTML,CSS,JS and NPM modules (to make the chessboard) or with some WASM-compiled code that puts Graphviz (the application that provides the graph generator `dot`) inside Pulsar, like this code that inspects an app state and draws a map of dependencies:
![DOT](https://user-images.githubusercontent.com/138037/193501974-239089b2-9ae9-4783-b52f-d1738c9aed1b.gif)

Are these possible on VSCode? Partially. Some are, but are slower because we need to pass messages between the "editor part" and the "webview part". The Graphviz, and anything that uses NPM modules are actually out of reach.

It's also _way easier_ on Pulsar - I don't need any kind of "message passing" or "registering webviews" and such - it's just Electron.

**Reply by Daeraxa** [reply]
This is fascinating, I didn't realise just how far it could actually go and it makes total sense to me that this is the area we need to hit hardest in terms of "marketing" Pulsar.
This is exactly the kind of stuff I think we need to showcase on the website, Pulsar can be everything from a simple text editor to doing insane things like this with.
